# Building Logistic Regression Model for Absenteeism Project

Now that we have our preprocessed data, we can start building a model that can take in the predictors and predict whether a random person is likely to be absent at work.

In this notebook, we will be creating a Logistic Regression (Logit) model to do the job.

## Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np

## Import the Data

In [2]:
data_preprocessed = pd.read_csv('../Dataset/Absenteeism_preprocessed.csv')
data_preprocessed

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          0         2     1                          4  
1                 31          0         1     0                          0  
2                 31          0         0     0                          2  
3                 24          0         2     0                          4  
4                 30          0         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          1         2     0                          8  
696               24          0         1     2                          3  
697               25          1         0     0                          8  
698               25          1         0     0                          2  
699               25          0         1     1                          2  

[700 rows x 15 columns]

## Create the Targets

For this project, we will be differentiating between excessive absenteeism and moderate absenteeism. For that, we will create target variable containing 1 or 0 for each depending on if the `Absenteeism Time in Hours` is more than or less than its median value respectively.

Finally, we can drop the `Absenteeism Time in Hours` column from the dataframe as it won't be needed anymore.

In [4]:
# create a target variable based on the median of `Absenteeism Time in Hours`
targets = np.where(data_preprocessed['Absenteeism Time in Hours'] > data_preprocessed['Absenteeism Time in Hours'].median(), 1, 0)
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [5]:
# add the newly created targets to the dataframe
data_preprocessed['Excessive Absenteeism'] = targets
data_preprocessed

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          0         2     1                          4   
1                 31          0         1     0                          0   
2                 31          0         0     0                          2   
3                 24          0         2     0                          4   
4                 30          0         2     1                          2   
..               ...        ...       ...   ...                        ...   
695               22          1         2     0                          8   
696               24          0         1     2                          3   
697               25          1         0     0                          8   
698               25          1         0     0                          2   
699               25          0         1     1                          2   

     Excessive Absenteeism  
0                        1  
1                        0  
2                        0  
3                        1  
4                        0  
..                     ...  
695                      1  
696                      0  
697                      1  
698                      0  
699                      0  

[700 rows x 16 columns]

In [6]:
# drop the `Absenteeism in Hours` column
data_with_targets = data_preprocessed.drop(['Absenteeism Time in Hours'], axis=1)

data_with_targets

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0                 30          0         2     1                      1  
1                 31          0         1     0                      0  
2                 31          0         0     0                      0  
3                 24          0         2     0                      1  
4                 30          0         2     1                      0  
..               ...        ...       ...   ...                    ...  
695               22          1         2     0                      1  
696               24          0         1     2                      0  
697               25          1         0     0                      1  
698               25          1         0     0                      0  
699               25          0         1     1                      0  

[700 rows x 15 columns]

### Balancing classes

For a good classification model, it is necessary to balance the classes of the samples in the dataset. It will create more generalised model.

In [7]:
# check the proportion of the 2 classes
targets.sum() / targets.shape[0]

0.45571428571428574

As we can see, around 46% of the data belongs to class 1 (Excessive Absenteeism) while the rest 54% to class 0 (Moderate Absenteeism). Hence, the dataset is roughly balanced and does not need further interference.